### PR TITLE
Add the needed APIs for Skia

### PR DESCRIPTION
**Pull request by younies**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -116,6 +122,15 @@ pub struct ParagraphInfo {
     pub level: Level,
 }
 
+
+
+impl ParagraphInfo {
+    /// Get the length of the paragraph in the source text.
+    pub fn get_length(&self) -> usize {
```
Comment (Manishearth) [inline src/lib.rs:129]: nit: just `len()`

File: src/lib.rs [hunk 2]
```diff
@@ -348,6 +363,36 @@ impl<'text> BidiInfo<'text> {
...
+    //     if even {
+    //         return Direction::LTR;
+    //     }
+
+    //     Direction::RTL
+    // } 
+
+
+    /// Returns the `Level` of a certain character in the paragraph.
+    pub fn get_level_at(&self, para: &ParagraphInfo, position: usize) -> Level {
```
Comment (Manishearth) [inline src/lib.rs:391]: nit: just `level_at()` i think

File: src/lib.rs [hunk 3]
```diff
@@ -348,6 +363,36 @@ impl<'text> BidiInfo<'text> {
         }
     }
 
+    // /// Returns if the paragraph is Left direction, right direction or mixed.
+    // pub fn get_direction(&self, para: &ParagraphInfo) -> Direction  {
```
Comment (mbrubeck) [inline src/lib.rs:367]: Please remove the commented-out code.

**Review by mbrubeck — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 3 above)
Comment (Manishearth) [reply src/lib.rs:367]: (I think this is just part of their WIP work, it's a draft PR)

Comment (mbrubeck) [reply src/lib.rs:367]: 👍

Comment (younies) [inline src/lib.rs:367]: @younies , document all the behaviour and comment.

File: src/lib.rs [hunk 4]
```diff
@@ -348,6 +363,36 @@ impl<'text> BidiInfo<'text> {
         }
     }
 
+    // /// Returns if the paragraph is Left direction, right direction or mixed.
```
Comment (Manishearth) [inline src/lib.rs:366]: document
 - behavior for empty paras
 - behavior for paras that don't come from this object

File: src/lib.rs [hunk 5]
```diff
@@ -348,6 +363,36 @@ impl<'text> BidiInfo<'text> {
...
+    // pub fn get_direction(&self, para: &ParagraphInfo) -> Direction  {
+    //     let mut even = false;
+    //     let mut odd = false;
+    //     for i in para.range {
+    //         if self.levels[i].into()%2  == 0 {
```
Comment (Manishearth) [inline src/lib.rs:371]: just use `.is_ltr()` or `.is_rtl()`

**Comment by Manishearth** [comment]
A lot of these APIs take in an `&ParagraphInfo` and what we might do is build a `struct Paragraph<'a>` that contains both a BidiInfo and ParagraphInfo and add methods _there_

File: src/lib.rs [hunk 6]
```diff
@@ -102,6 +102,12 @@ use core::ops::Range;
 use crate::BidiClass::*;
 use crate::format_chars as chars;
 
+pub enum Direction {
```
[the remark this answers is not in the document]
Comment (younies) [reply src/lib.rs:106]: done!

File: src/lib.rs (hunk 1 above)
Comment (younies) [reply src/lib.rs]: done in PR: https://github.com/servo/unicode-bidi/pull/73

File: src/lib.rs (hunk 4 above)
Comment (younies) [reply src/lib.rs]: for empty para, done in the tests! shall we document?

File: src/lib.rs (hunk 5 above)
Comment (younies) [reply src/lib.rs]: done!

File: src/lib.rs [hunk 7]
```diff
@@ -516,6 +524,53 @@ impl<'text> BidiInfo<'text> {
...
+
+    pub fn new(info: &'a BidiInfo<'text>, para : &'a ParagraphInfo) ->  Paragraph<'a, 'text> {
+        Paragraph{info,para}
+    }  
+
+    /// Returns if the paragraph is Left direction, right direction or mixed. 
+    pub fn direction(&self) ->  Direction {
+        let mut ltr = false;
+        let mut rtl = false;
+        for i in self.para.range.start..self.para.range.end  {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply src/lib.rs:545]: It's because ranges aren't Copy, but you can call `range.clone()` cheaply.

File: src/lib.rs [hunk 8]
```diff
@@ -561,6 +524,53 @@ impl<'text> BidiInfo<'text> {
     }
 }
 
+
+/// Contains a reference of `Bidi_info` and one of its `paragraphs`.
```
Comment (Manishearth) [inline src/lib.rs:528]: Nit: I think we need more docs here.

File: src/lib.rs [hunk 9]
```diff
@@ -561,6 +524,53 @@ impl<'text> BidiInfo<'text> {
...
+#[derive(Debug)]
+pub struct Paragraph<'a, 'text> {
+  info: &'a BidiInfo<'text>,
+  para: &'a ParagraphInfo,
+} 
+
+impl <'a, 'text>Paragraph <'a, 'text>{
```
Comment (Manishearth) [inline src/lib.rs:535]: We should have ways to construct this from an iterator. I can add them if you'd like since iterators are a bit tricky.

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 10]
```diff
@@ -561,6 +524,53 @@ impl<'text> BidiInfo<'text> {
     }
 }
 
+
+/// Contains a reference of `Bidi_info` and one of its `paragraphs`.
+#[derive(Debug)]
+pub struct Paragraph<'a, 'text> {
+  info: &'a BidiInfo<'text>,
```
[the remark this answers is not in the document]
Comment (younies) [reply src/lib.rs:531]: done.

File: src/lib.rs (hunk 8 above)
Comment (younies) [reply src/lib.rs:528]: Do you mean add a usage? or illustrate more?

File: src/lib.rs (hunk 9 above)
Comment (younies) [reply src/lib.rs:535]: If it is okay, I can add this in a following PR.

File: src/lib.rs [hunk 11]
```diff
@@ -561,6 +524,53 @@ impl<'text> BidiInfo<'text> {
...
+  para: &'a ParagraphInfo,
+} 
+
+impl <'a, 'text>Paragraph <'a, 'text>{
+
+    pub fn new(info: &'a BidiInfo<'text>, para : &'a ParagraphInfo) ->  Paragraph<'a, 'text> {
```
[the remark this answers is not in the document]
Comment (younies) [reply src/lib.rs:537]: done, but I used `cargo fmt`, why those formatting did not formatted?

File: src/lib.rs [hunk 12]
```diff
@@ -561,6 +524,53 @@ impl<'text> BidiInfo<'text> {
...
+        Paragraph{info,para}
```
[the remark this answers is not in the document]
Comment (younies) [reply src/lib.rs:538]: done

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (younies) [reply src/lib.rs:545]: thx, done.

File: src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply src/lib.rs]: I fixed cargo fmt just now

File: src/lib.rs (hunk 8 above)
Comment (Manishearth) [reply src/lib.rs:533]: Just illustrate more what this is for and why it exists.

File: src/lib.rs (hunk 9 above)
Comment (Manishearth) [reply src/lib.rs]: okay

File: src/lib.rs (hunk 8 above)
Comment (younies) [reply src/lib.rs:533]: done

**Pull request merged**